src/MongoDB.Driver/PipelineStageDefinitionBuilder.cs

```diff
@@ -927,6 +927,108 @@ public static class PipelineStageDefinitionBuilder
+                }
+            }
+            else
+            {
+                if (mergeOptions.WhenMatched != MergeStageWhenMatched.Pipeline)
+                {
+                    throw new ArgumentException("WhenMatchedPipeline can only be set when WhenMatched == Pipeline.");
+                }
+            }
+
+            const string operatorName = "$merge";
+            var stage = new DelegatedPipelineStageDefinition<TInput, TOutput>(
+                operatorName,
+                (inputSerializer, serializerRegistry) =>
+                {
+                    var outputSerializer = mergeOptions.OutputSerializer ?? (inputSerializer as IBsonSerializer<TOutput>) ?? serializerRegistry.GetSerializer<TOutput>();
+
+                    var outputCollectionNamespace = outputCollection.CollectionNamespace;
+                    var outputDatabaseNamespace = outputCollectionNamespace.DatabaseNamespace;
+                    var renderedInto = new BsonDocument
+                    {
+                        { "db", outputDatabaseNamespace.DatabaseName },
+                        { "coll", outputCollectionNamespace.CollectionName }
+                    };
+
+                    BsonValue renderedOn = null;
+                    if (mergeOptions.OnFieldNames != null)
+                    {
+                        if (mergeOptions.OnFieldNames.Count == 1)
+                        {
+                            renderedOn = mergeOptions.OnFieldNames.Single();
+                        }
+                        else
+                        {
+                            renderedOn = new BsonArray(mergeOptions.OnFieldNames.Select(n => BsonString.Create(n)));
+                        }
+                    }
+
+                    BsonValue renderedWhenMatched = null;
+                    if (mergeOptions.WhenMatched.HasValue)
+                    {
+                        var whenMatched = mergeOptions.WhenMatched.Value;
+                        if (whenMatched == MergeStageWhenMatched.Pipeline)
+                        {
+                            var renderedPipeline = mergeOptions.WhenMatchedPipeline.Render(outputSerializer, serializerRegistry);
+                            renderedWhenMatched = new BsonArray(renderedPipeline.Documents);
+                        }
+                        else
+                        {
+                            renderedWhenMatched = MongoUtils.ToCamelCase(whenMatched.ToString());
+                        }
+                    }
+
+                    BsonString renderedWhenNotMatched = null;
+                    if (mergeOptions.WhenNotMatched.HasValue)
+                    {
+                        var whenNotMatched = mergeOptions.WhenNotMatched;
+                        renderedWhenNotMatched = MongoUtils.ToCamelCase(whenNotMatched.ToString());
+                    }
+
+                    var renderedMerge = new BsonDocument
+                    {
```

Comment:
Probably we need to add tests to validate the below options?

Reply:
Coming in a later commit.

Reply:
Add in the most recent commit.